### PR TITLE
[All] Factorize and rename TopologyObjectType into TopologyElementType

### PR DESCRIPTION
**Pull request by epernod**
- Rename ```TopologyObjectType```  into ```TopologyElementType``` (this enum is used to define the type of upper element contain in the current topology)
- Factorize methods and member storing this information into ```BaseMeshTopology::m_upperElementType``` 
- Update ```DiagonalMass``` and ```MeshMatrixMass``` which were using their own enum to store the type of topology. 

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][force-full-build]

**Comment by epernod** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaCore/src/sofa/core/topology/Topology.h [hunk 1]
```diff
@@ -37,21 +37,22 @@ namespace topology
 {
 
 /// The enumeration used to give unique identifiers to Topological objects.
-enum TopologyObjectType
+enum TopologyElementType
```
Comment (fredroy) [inline SofaKernel/modules/SofaCore/src/sofa/core/topology/Topology.h]: [**enum class**](https://www.stroustrup.com/C++11FAQ.html#enum) should be preferred when it is possible

**Review by fredroy — CHANGES_REQUESTED** [review]

File: SofaKernel/modules/SofaCore/src/sofa/core/topology/Topology.h (hunk 1 above)
Comment (epernod) [reply SofaKernel/modules/SofaCore/src/sofa/core/topology/Topology.h]: yes but... if I do that change I get 11937 errors....

**Review by fredroy — APPROVED** [review]

**Comment by fredroy** [comment]
Last thing before setting ready 🙃,  dont you think it would be a good idea setting the previous enum with the attribute like [[deprecated "Change to TopologyElementType PR1593"]]? 
It would inform users (who did not read the Changelog) where to look...

https://en.cppreference.com/w/cpp/language/attributes/deprecated
- enumeration: enum [[deprecated]] E {};,

**Comment by epernod** [comment]
yes good idea, but just a deprecated info? or a nice and smooth compatibility transition like: 
warning and convert old enum into new enum? ^^

**Comment by fredroy** [comment]
> 
> 
> yes good idea, but just a deprecated info? or a nice and smooth compatibility transition like:
> warning and convert old enum into new enum? ^^

I guess deprecated attribute + the old enum values with the new ones would do the trick no ?

**Pull request merged**